psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,33 +788,798 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                        int Q = task_shells[Q2start + Q2];
+                        int S = task_shells[S2start + S2];
+                        int Qsize = primary_->shell(Q).nfunction();
+                        int Ssize = primary_->shell(S).nfunction();
+                        int Qoff = primary_->shell(Q).function_index();
+                        int Soff = primary_->shell(S).function_index();
+                        int Qoff2 = task_offsets[Q2 + Q2start] - task_offsets[Q2start];
+                        int Soff2 = task_offsets[S2 + S2start] - task_offsets[S2start];
+                        for (int q = 0; q < Qsize; q++) {
+                            for (int s = 0; s < Ssize; s++) {
+#pragma omp atomic
+                                Kp[q + Qoff][s + Soff] += K4p[(q + Qoff2) * dSsize + s + Soff2];
+                                if (!lr_symmetric_) {
+#pragma omp atomic
+                                    Kp[s + Soff][q + Qoff] += K8p[(s + Soff2) * dQsize + q + Qoff2];
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+        }  // End stripe out
+        // if (thread == 0) timer_off("JK: Atomic");
+
+    }  // End master task list
+
+    if (build_J) {
+        for (auto Jmat : J) {
+            Jmat->scale(2.0);
+            Jmat->hermitivitize();
+        }
+    }
+
+    if (build_K && lr_symmetric_) {
+        for (auto Kmat : K) {
+            Kmat->scale(2.0);
+            Kmat->hermitivitize();
+        }
+    }
+
+    if (bench_) {
+        auto mode = std::ostream::app;
+        auto printer = PsiOutStream("bench.dat", mode);
+        size_t ntri = nshell * (nshell + 1L) / 2L;
+        size_t possible_shells = ntri * (ntri + 1L) / 2L;
+        printer.Printf("Computed %20zu Shell Quartets out of %20zu, (%11.3E ratio)\n", computed_shells,
+                        possible_shells, computed_shells / (double)possible_shells);
+    }
+
+    for (int i = 0; i < computed_integrals_.size(); i++) {
+        computed_integrals_[i].clear();
+    }
+
+    timer_off("build_JK_matrices()");
+}
+
+bool linK_sort_helper(const std::tuple<int, double>& t1, const std::tuple<int, double>& t2) {
+    return std::get<1>(t1) > std::get<1>(t2);
+}
+
+void DirectJK::build_linK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, const std::vector<SharedMatrix>& D,
+                  std::vector<SharedMatrix>& J, std::vector<SharedMatrix>& K) {
+
+    timer_on("build_linK()");
+
+    // => Zeroing <= //
+
+    for (auto Jmat : J) {
+        Jmat->zero();
+    }
+
+    for (auto Kmat : K) {
+        Kmat->zero();
+    }
+
+    // => Sizing <= //
+
+    int nshell = primary_->nshell();
+    int nthread = df_ints_num_threads_;
+
+    // => Setting up computed integrals map <= //
+    computed_integrals_.resize(nthread);
+
+    // => Task Blocking <= //
+
+    std::vector<int> task_shells;
+    std::vector<int> task_starts;
```

Comment:
I understand that you're just reusing old code, but old code can be badly written. Please rename to something more descriptive, like `atom_nshells` and `atom_first_shell_idx`.